src/torchcodec/samplers/_implem.py

```diff
@@ -323,3 +331,236 @@ def clips_at_regular_indices(
+    for start_seconds in clip_start_seconds:
+        clip_timestamps = [
+            start_seconds + i * seconds_between_frames
+            for i in range(num_frames_per_clip)
+        ]
+        clip_timestamps = [
+            timestamp for timestamp in clip_timestamps if timestamp < end_stream_seconds
+        ]
+
+        if len(clip_timestamps) < num_frames_per_clip:
+            clip_timestamps = policy_fun(clip_timestamps, num_frames_per_clip)
+        all_clips_timestamps += clip_timestamps
+
+    return all_clips_timestamps
+
+
+def _decode_all_clips_timestamps(
+    decoder: VideoDecoder, all_clips_timestamps: list[float], num_frames_per_clip: int
+) -> list[FrameBatch]:
+    # This is 99% the same as _decode_all_clips_indices. The only change is the
+    # call to .get_frame_displayed_at(pts) instead of .get_frame_at(idx)
+
+    def chunk_list(lst, chunk_size):
+        # return list of sublists of length chunk_size
+        return [lst[i : i + chunk_size] for i in range(0, len(lst), chunk_size)]
+
+    def to_framebatch(frames: list[Frame]) -> FrameBatch:
+        # IMPORTANT: see other IMPORTANT note below
+        data = torch.stack([frame.data for frame in frames])
+        pts_seconds = torch.tensor([frame.pts_seconds for frame in frames])
+        duration_seconds = torch.tensor([frame.duration_seconds for frame in frames])
+        return FrameBatch(
+            data=data, pts_seconds=pts_seconds, duration_seconds=duration_seconds
+        )
+
+    all_clips_timestamps_sorted, argsort = zip(
+        *sorted(
+            (frame_index, i) for (i, frame_index) in enumerate(all_clips_timestamps)
+        )
+    )
+    previous_decoded_frame = None
+    all_decoded_frames = [None] * len(all_clips_timestamps)
+    for i, j in enumerate(argsort):
+        frame_pts_seconds = all_clips_timestamps_sorted[i]
+        if (
+            previous_decoded_frame is not None  # then we know i > 0
+            and frame_pts_seconds == all_clips_timestamps_sorted[i - 1]
+        ):
+            # Avoid decoding the same frame twice.
+            # IMPORTANT: this is only correct because a copy of the frame will
+            # happen within `to_framebatch` when we call torch.stack.
+            # If a copy isn't made, the same underlying memory will be used for
+            # the 2 consecutive frames. When we re-write this, we should make
+            # sure to explicitly copy the data.
+            decoded_frame = previous_decoded_frame
+        else:
+            decoded_frame = decoder.get_frame_displayed_at(seconds=frame_pts_seconds)
+        previous_decoded_frame = decoded_frame
+        all_decoded_frames[j] = decoded_frame
+
+    all_clips: list[list[Frame]] = chunk_list(
+        all_decoded_frames, chunk_size=num_frames_per_clip
+    )
+
+    return [to_framebatch(clip) for clip in all_clips]
+
+
+def clips_at_regular_timestamps(
+    decoder,
+    *,
+    seconds_between_clip_starts: float,
```

Comment:
The user will have to do some math to arrive at a number here that ensures the full video is used and we don't apply a policy function, right?

Thinking about this a bit more, I think this may be easy to get wrong for users and it is not consistent with `clilps_at_regular_indices` which takes in a num_clips.

For 2 clips, a naive user may just use seconds_between_clip_starts = video_duration (or video_duration / 2)

For 3 clips a naive user may use seconds_between_clip_starts = video_duration / 2 (or video_duration/3).

It seems easy to get wrong (off by one), etc. And doesn't offer an easy way to transition between index-based and time-based (params have to be changed).

I think the params here should be:

```
    num_clips: int = 1,
    num_frames_per_clip: int = 1,
    seconds_between_clip_frames: float = 1,  # This is the equivalent of indices_between_clip_frames
```

With this construction we probably don't even need a policy function if we exclude the upper bound properly like torch.arange, etc. (We may need a policy function if we want to sample all frames with equal probability).

Reply:
I don't find `seconds_between_clips` particularly intuitive either.

I'm happy to introduce a `num_clips` parameter that would  be mutually exclusive with `seconds_between_clip_starts`.
 However, as discussed during the design phase, we'll have to keep `seconds_between_clip_starts` for feature-parity with existing samples.

Reply:
I think some guidance on how to derive this seconds_between_clip_starts should be useful as a comment or function since I do think it may be easy to get wrong. Add it to the test as well, so users can see how it's used.

Reply:
I don't know enough about python to see how mutually exclusive args would be constructed, but num_clips seems better for users who want to transition from fixed FPS to variable FPS videos.

In C++ one solution may be to do function overloading where the same name either accepts num_clips or seconds_between_clip_starts as two different types. I don't know enough python to know what's pythonic acceptable in python.